osu.Game/Database/LegacyModelExporter.cs

```diff
@@ -0,0 +1,164 @@
+        {
+            // check if the model is being exporting already
+            if (!exportingModels.Contains(model))
+            {
+                exportingModels.Add(model);
+            }
+            else
+            {
+                PostNotification?.Invoke(new SimpleErrorNotification
+                {
+                    Text = "File is being exported"
+                });
+                return;
+            }
+
+            string itemFilename = GetFilename(model).GetValidFilename();
+            IEnumerable<string> existingExports =
+                exportStorage
+                    .GetFiles(string.Empty, $"{itemFilename}*{FileExtension}")
+                    .Concat(exportStorage.GetDirectories(string.Empty));
+            string filename = NamingUtils.GetNextBestFilename(existingExports, $"{itemFilename}{FileExtension}");
+            bool success = false;
+
+            ProgressNotification notification = new ProgressNotification
+            {
+                State = ProgressNotificationState.Active,
+                Text = "Exporting...",
+                CompletionText = "Export completed"
+            };
+            notification.CompletionClickAction += () => exportStorage.PresentFileExternally(filename);
+            PostNotification?.Invoke(notification);
+
+            try
+            {
+                using (var stream = exportStorage.CreateFileSafely(filename))
+                {
+                    success = await ExportToStreamAsync(model, stream, notification, cancellationToken ?? notification.CancellationToken).ConfigureAwait(false);
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                success = false;
+            }
+            finally
+            {
+                // cleanup if export is failed or canceled.
+                if (!success)
+                {
+                    exportStorage.Delete(filename);
+                }
+
+                exportingModels.Remove(model);
+            }
+        }
+
+        /// <summary>
+        /// Export model to stream.
+        /// </summary>
+        /// <param name="model">The medel which have <see cref="IHasGuidPrimaryKey"/>.</param>
+        /// <param name="stream">The stream to export.</param>
+        /// <param name="notification">The notification will displayed to the user</param>
+        /// <param name="cancellationToken">The Cancellation token that can cancel the exporting.</param>
+        /// <returns>Whether the export was successful</returns>
+        public async Task<bool> ExportToStreamAsync(TModel model, Stream stream, ProgressNotification? notification = null, CancellationToken cancellationToken = default)
+        {
+            ProgressNotification notify = notification ?? new ProgressNotification();
+
+            Guid id = model.ID;
+            return await Task.Run(() =>
```

Comment:
I'm confused. Why are you using `Task.Run` here? You're in an `async` method. Recommendation would be

```diff
diff --git a/osu.Game/Database/LegacyModelExporter.cs b/osu.Game/Database/LegacyModelExporter.cs
index acb83ce1d5..03f4c06297 100644
--- a/osu.Game/Database/LegacyModelExporter.cs
+++ b/osu.Game/Database/LegacyModelExporter.cs
@@ -124,12 +124,13 @@ public async Task ExportAsync(TModel model, CancellationToken? cancellationToken
         /// <param name="notification">The notification will displayed to the user</param>
         /// <param name="cancellationToken">The Cancellation token that can cancel the exporting.</param>
         /// <returns>Whether the export was successful</returns>
-        public async Task<bool> ExportToStreamAsync(TModel model, Stream stream, ProgressNotification? notification = null, CancellationToken cancellationToken = default)
+        public Task<bool> ExportToStreamAsync(TModel model, Stream stream, ProgressNotification? notification = null, CancellationToken cancellationToken = default)
         {
             ProgressNotification notify = notification ?? new ProgressNotification();
 
             Guid id = model.ID;
-            return await Task.Run(() =>
+
+            return Task.Run(() =>
             {
                 realmAccess.Run(r =>
                 {
@@ -148,7 +149,7 @@ public async Task<bool> ExportToStreamAsync(TModel model, Stream stream, Progres
                 notify.CompletionText = "Export Complete, Click to open the folder";
                 notify.State = ProgressNotificationState.Completed;
                 return true;
-            }, cancellationToken).ConfigureAwait(false);
+            }, cancellationToken);
         }
 
         /// <summary>

```